File: Data Structure/Tree/Binary Tree/Binary Search Tree/Python3/108_convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.tree_helper(nums, 0, len(nums) - 1)
        
        
    def tree_helper(self, nums, leftIndex, rightIndex) -> Optional[TreeNode]:
        if leftIndex > rightIndex:
            return None
        
        if leftIndex == rightIndex:
            return TreeNode(nums[leftIndex])
        
        position = leftIndex + (rightIndex - leftIndex) // 2
        root = TreeNode(nums[position])
        left = self.tree_helper(nums, leftIndex, position - 1)
        right = self.tree_helper(nums, position + 1, rightIndex)
        
        root.left = left
        root.right = right
        
        return root

File: Data Structure/Tree/Binary Tree/Binary Search Tree/Java/108_convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return tree_helper(nums, 0, nums.length - 1);   
    }
    
    public TreeNode tree_helper(int[] nums, int leftIndex, int rightIndex) {
        if (leftIndex > rightIndex) {
            return null;
        }
        
        if (leftIndex == rightIndex) {
            return new TreeNode(nums[leftIndex]);
        }
        
        int position = leftIndex + (rightIndex - leftIndex) / 2;
        
        TreeNode root = new TreeNode(nums[position]);
        TreeNode left = tree_helper(nums, leftIndex, position - 1);
        TreeNode right = tree_helper(nums, position + 1, rightIndex);
        
        root.left = left;
        root.right = right;
        
        return root;
    }
}